escape_roomba/thread_manager.py

```diff
@@ -77,261 +82,421 @@ async def _on_ready(self):
+            return  # Thread was removed at some point.
+
+        old_len = len(thread.intro_messages or [])
+        thread.intro_messages = [
+            m async for m in thread.thread_channel.history(
+                limit=self._FETCH_INTRO, oldest_first=True)]
+        if self._logger.isEnabledFor(logging.DEBUG):
+            ims = thread.intro_messages
+            self._logger.debug(
+                f'Fetched #{thread.thread_channel.name} intro '
+                f'({old_len} => {len(ims)}m):' +
+                ''.join(f'\n    {fobj(m=m)}' for m in ims))
+
+        # When the intro *shrinks*, re-check the origin in case the
+        # thread channel is now eligible for removal.
+        if len(thread.intro_messages) < old_len:
+            await self._async_fetch_locked_message(
+                thread.origin_channel_id, thread.origin_message_id)
 
-            if t is not old_t:
-                if self._logger.isEnabledFor(logging.DEBUG):
-                    self._logger.debug(f'Found #{channel.name}:\n'
-                                       f'    oc={fid(ci)} om={fid(mi)}\n'
-                                       f'    tc={fid(channel.id)}\n'
-                                       f'    topic="{channel.topic}"')
+    async def _async_fetch_recents(self, channel):
+        """Fetches and examines the recent history of a channel that appeared
+        (startup, server join, channel creation, visibility change)."""
 
-                # For a new channel, update the origin message.
-                # TODO - avoid redundancy if we just created the thread?
-                await self._async_maybe_update(channel_id=ci, message_id=mi)
+        if channel.type != discord.ChannelType.text:
+            return
+
+        # Go through the history and trigger updates for unprocessed messages;
+        # avoid re-fetching any message that is already a thread origin.
+        await asyncio.gather(*[
+            self._async_check_message(message=m)
+            async for m in channel.history(
+                limit=self._FETCH_RECENT, oldest_first=False)
+            if m.id not in self._thread_by_origin.get(channel.id, {})])
 
     async def _async_forget_channel(self, channel):
         """Handles a channel that is no longer visible (server detach,
         channel deleted, visibility change)."""
 
         # If it was a thread channel, remove it from tracking.
-        t = self._thread_by_channel.pop(channel.id, None)
+        t = self._thread_by_channel.get(channel.id)
         if t is not None:
-            ci, mi = t.origin_channel_id, t.origin_message_id
-            del self._thread_by_origin[ci][mi]
-            if self._logger.isEnabledFor(logging.DEBUG):
-                self._logger.debug(f'Removed #{channel.name}:\n'
-                                   f'    oc={fid(ci)} om={fid(mi)}\n'
-                                   f'    tc={fid(channel.id)}')
+            # Lock and check again to avoid conflicts.
```

Comment:
I don't fully follow the logic here. Why not check with a lock to begin with? (i.e. why do the first non-locked check at all?) (Are you doing it as a perf optimization? If so, is that really worth it vs simplifying the code?)

Reply:
Reorganized and clarified a bit. We need to look up the _Thread to even know the message ID to lock, but while waiting for the lock other changes might happen and the _Thread could already be removed (I actually ran into this in testing, was getting errors from the 'del').